### PR TITLE
Enable committing offsets in kafka to gcs, BQ template

### PR DESCRIPTION
**Pull request by AnandInguva**

### CODE REVIEW

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java [hunk 1]
```diff
@@ -49,23 +49,49 @@ public interface KafkaReadOptions extends PipelineOptions {
 
   void setKafkaReadTopics(String inputTopics);
 
-  @TemplateParameter.Enum(
+  @TemplateParameter.Boolean(
       order = 3,
       groupName = "Source",
+      name = "enableCommitOffSets",
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:55]: "offset" is a single work so the parameter name should be "enableCommitOffsets" (OffSets -> Offsets).

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java [hunk 2]
```diff
@@ -49,23 +49,49 @@ public interface KafkaReadOptions extends PipelineOptions {
...
+      optional = true,
+      description = "Enable committing offsets to Kafka",
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:57]: I suggest renaming this to "Commit Offsets to Kafka" (keep labels short and capitalized).

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java [hunk 3]
```diff
@@ -49,23 +49,49 @@ public interface KafkaReadOptions extends PipelineOptions {
...
+      helpText =
+          "Enable KafkaIO to commit offsets after consuming the record. Enabling this requires you to provide"
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:59]: I suggest changing the help text to:

"Commit offsets of processed messages to Kafka. If enabled, this will minimize the gaps or duplicate processing of messages when restarting the pipeline. Requires specifying the Consumer Group ID."

because:

1) Users of the template may not know what "KafkaIO" means, better to not mention it.
2) "or the template launcher will fail" - an unnecessary detail.
3) Added some more details to explain why this may be useful ("minimize the gaps ..." is copied from KafkaIO javadoc).

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java [hunk 4]
```diff
@@ -49,23 +49,49 @@ public interface KafkaReadOptions extends PipelineOptions {
...
+  void setEnableCommitOffSets(Boolean value);
+
+  @TemplateParameter.Text(
+      order = 4,
+      groupName = "Source",
+      optional = true,
+      description = "Consumer group id to commit offsets",
+      helpText = "Consumer group ID to commit offsets to Kafka.")
+  @Default.String("")
+  String getConsumerGroupID();
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:73]: `ID` -> `Id` as per the code style.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java [hunk 5]
```diff
@@ -49,23 +49,49 @@ public interface KafkaReadOptions extends PipelineOptions {
...
+  @Default.Boolean(false)
+  Boolean getEnableCommitOffSets();
+
+  void setEnableCommitOffSets(Boolean value);
+
+  @TemplateParameter.Text(
+      order = 4,
+      groupName = "Source",
+      optional = true,
+      description = "Consumer group id to commit offsets",
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:70]: "Consumer group id to commit offsets" -> "Consumer Group ID" (keep it short and capitalized).

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java [hunk 6]
```diff
@@ -162,4 +164,21 @@ public static Map<String, Object> configureKafka(Map<String, String> props) {
     }
     return config;
   }
+
+  public static Map<String, Object> configureKafkaOffsetCommit(KafkaReadOptions options) {
+    Map<String, Object> kafkaConfig = new HashMap<>();
+    if (options.getEnableCommitOffSets() && options.getConsumerGroupID().isBlank()) {
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java:170]: I think `options.getConsumerGroupID()` can be null? If so, `.isBlank()` can cause a NPE.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 2 above)
Comment (Dippatel98) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:57]: nit: Change to `Enable Commit Offsets`

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 5 above)
Comment (Dippatel98) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java:70]: nit: Change to `Consumer Group ID`

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java [hunk 7]
```diff
@@ -162,4 +164,21 @@ public static Map<String, Object> configureKafka(Map<String, String> props) {
...
+      throw new IllegalArgumentException(
+          "Committing offsets is enabled in the template but a consumer group ID is not provided."
+              + "Offset management requires a consumer group ID.");
+    }
+
+    kafkaConfig.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, options.getKafkaReadOffset());
+
+    // Set the Kafka Consumer Group ID
+    if (!options.getConsumerGroupID().isBlank()) {
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java:179]: Same here.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/KafkaTransform.java [hunk 8]
```diff
@@ -108,11 +107,12 @@ public static PTransform<PBegin, PCollection<KV<byte[], GenericRecord>>> readAvr
    * @param config configuration for the Kafka consumer
    * @return PCollection of Kafka Key & Value Pair deserialized in string format
    */
-  public static PTransform<PBegin, PCollection<KafkaRecord<byte[], byte[]>>> readBytesFromKafka(
+  public static KafkaIO.Read<byte[], byte[]> readBytesFromKafka(
```
Comment (Dippatel98) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/KafkaTransform.java:110]: Just curious, why was this changed?

**Review by an2x — CHANGES_REQUESTED** [review]

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java [hunk 9]
```diff
@@ -206,6 +207,7 @@ public static PipelineResult run(KafkaToBigQueryFlexOptions options)
           ImmutableMap.<String, Object>builder()
               .putAll(kafkaConfig)
               .putAll(setClientAuthConfig(username, password))
+              .putAll(KafkaCommonUtils.configureKafkaOffsetCommit(options))
```
Comment (Dippatel98) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java:210]: I don't think this is the correct place for this. This will only happen if the authentication mode is `SASL_PLAIN`. This should be added outside the if statement.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/KafkaTransform.java (hunk 8 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/KafkaTransform.java:110]: I thought this would be a better type annotation than `PTransforn....`, just for readability

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 1 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Done

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 2 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Done

Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Changed it to `Commit Offsets to Kafka`

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 3 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Yep. this sounds good to me. Thanks

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 5 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Done

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java (hunk 4 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/options/KafkaReadOptions.java]: Done

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java (hunk 6 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java]: I provide a default value - `""`, which is a blank value right?

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java (hunk 9 above)
Comment (AnandInguva) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java]: I moved it a different code path. PTAL @Dippatel98

File: v2/kafka-to-gcs/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToGcsFlex.java [hunk 10]
```diff
@@ -210,28 +208,26 @@ public static PipelineResult run(KafkaToGcsOptions options) throws UnsupportedOp
...
-                    ByteArrayDeserializer.class, NullableCoder.of(ByteArrayCoder.of()))
-                .withConsumerConfigUpdates(kafkaConfig));
-
+    KafkaIO.Read<byte[], byte[]> kafkaTransform =
+        KafkaTransform.readBytesFromKafka(
+            options.getReadBootstrapServers(),
+            topics,
+            kafkaConfig,
+            null,
+            options.getEnableCommitOffSets());
```
Comment (AnandInguva) [inline v2/kafka-to-gcs/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToGcsFlex.java:229]:
```suggestion
            options.getEnableCommitOffsets());
```

**Review by an2x — APPROVED** [review]

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java (hunk 6 above)
Comment (an2x) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/utils/KafkaCommonUtils.java]: Yeah, this should work. Up to you if you want to add a stronger check here.
